# Exercise 4.6 Task

### This script contains the following points:
#### 1. (Step 3) Importing libraries, defining project path, and importing datasets
#### 2. (Step 4) Check dimensions of imported dataframes (df_ords_prods_combined and df_prods)
#### 3. (Step 5) Merge df_ords_combined_prods_combined_checked and df_prods_checked
#### 4. (Step 6) Confirming the merge with merge flag ('_2_merge')
#### 5. (Step 7) Exporting dataframe df_merged_large_2 as 'orders_products_merged.pkl'

## Step 3. Importing Libraries, Defining Project Path, and Importing Datasets

In [1]:
# Importing pandas, numpy, and os
import pandas as pd
import numpy as np
import os

In [2]:
# Defining project folder path
path = r'C:\Users\prena\03-2023 Instacart Basket Analysis'

In [3]:
# Importing orders_products_combined.pkl dataset
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
# Importing products_checked.csv dataset
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

## Step 4. Checking 'df_ords_prods_combined' and 'df_prods'

In [5]:
df_ords_prods_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   order_day_of_week  order_hour_of_day  days_since_last_order  product_id  \
0                  2                  8                    NaN         196   
1                  2                  8                    NaN       14084   
2                  2                  8                    NaN       12427   
3                  2                  8                    NaN       26088   
4                  2                  8                    NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [6]:
df_ords_prods_combined.shape

(32434489, 12)

#### Confirmed that the shape of 'df_ords_prods_combined' is the same shape as the dataframe exported from the script "4.6 Combining & exporting data PRE-WORK."

In [7]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [8]:
df_prods.shape

(49670, 6)

#### Confirmed that the shape of 'df_prods' is the same shape as the dataframe exported from the script "4.5 Data consistency checks"

#### Note: There are "Unnamed: ..." columns in both dataframes that I will remove here before merging:

In [9]:
# Dropping "Unnamed: 0.1" and "Unnamed: 0" from df_ords_prods_combined and renaming df_ords_prods_combined_checked

In [10]:
df_ords_prods_combined_checked = df_ords_prods_combined.drop(columns=["Unnamed: 0.1", "Unnamed: 0"])

In [11]:
df_ords_prods_combined_checked.shape

(32434489, 10)

In [12]:
df_ords_prods_combined_checked.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  
0                    NaN         196                  1          0   both  
1                    NaN       14084                  2          0   both  
2                    NaN       12427                  3          0   both  
3                    NaN       26088                  4          0   both  
4                    NaN       26405                  5          0   both

In [13]:
# Dropping "Unnamed: 0" from df_prods and renaming df_prods_checked

In [14]:
df_prods_checked = df_prods.drop(columns=["Unnamed: 0"])

In [15]:
df_prods_checked.shape

(49670, 5)

In [16]:
df_prods_checked.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## Step 5. Merge 'df_ords_prods_combined_checked' with 'df_prods_checked'
#### As mentioned in this exercise, for this project, we'll only be working with data sets that have a full merge rate. Looking at these two datasets, I suspect that there will be multiple orders that contain the same product.

In [17]:
# Inner join merge 'df_ords_prods_combined_checked' with 'df_prods_checked'
df_merged_large_2 = df_ords_prods_combined_checked.merge(df_prods_checked, on = 'product_id', indicator = '_merge_2')

In [18]:
df_merged_large_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  \
0                    NaN         196                  1          0   both   
1                   15.0         196                  1          1   both   
2                   21.0         196                  1          1   both   
3                   29.0         196                  1          1   both   
4                   28.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices _merge_2  
0         Soda        77              7     9.0     both  
1         Soda        77              7     9.0     both  
2         Soda        77              7     9.0     both  
3         Soda        77              7     9.0     both  
4         Soda        77              7     9.0     both

In [19]:
df_merged_large_2.shape

(32399732, 15)

#### The size of our rows went from 32,434,489 (in the larger df: df_ords_prods_combined_checked) to 32,399,732.
#### The size of or columns went from 10 (df_ords_prods_combined_checked) and 5 (df_prods_checked) to 15 (which makes sense, since the two dataframes share the column "product_id" and we added "_merge_2" as a merge flag).

## Step 6. Confirming the results of the merge using the merge flag.

In [20]:
# Check the frequency of the merge
df_merged_large_2['_merge_2'].value_counts()

both          32399732
left_only            0
right_only           0
Name: _merge_2, dtype: int64

## Step 7. Export df_merged_large_2 as 'orders_products_merged' as pkl file

In [21]:
# Exporting to pickle file 'orders_products_merged'
df_merged_large_2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))